### ISSUE TITLE
Accessibility

### ISSUE THREAD

**Issue by callumacrae** [NightCatSama/vue-slider-component#68]
Hey!

This component currently doesn't work with screen readers - because it's made of divs, the screen reader doesn't know it's an input element and goes straight over it. I'd recommend duplicating the state to a hidden range input for screenreaders to use.

Also, are you fine if I use this library as an example in a talk of an input library which doesn't work with screen readers? It 100% isn't intended as an attack and it won't be the only library I talk about - most libraries in the Vue ecosystem have the same problem!

**Comment by NightCatSama** [comment]
Because I do not have a screen readers, so it is difficult to reproduce the problem.

About add a hidden input element to compatibility, is refers to the events would only on the input trigger on a screen reader？

If yes, that would be very complicated to modify.

**Comment by callumacrae** [comment]
It wouldn't be tricky. Just add an `<input type="range">` slider with a [screen reader only](http://www.coolfields.co.uk/2016/05/text-for-screen-readers-only-updated/) class, and put `aria-hidden="true"` on the current HTML. Then you can bind the range slider to the variable with v-model and they would reflect each other's state.

**Comment by NightCatSama** [comment]
Whethe only need to add a hidden input(`<input type="range" aria-hidden="true" style="opacity: 0">`), and then synchronize the value of the component to the input value can be completed.

**Comment by callumacrae** [comment]
you want the aria-hidden attribute on the current HTML for vue-slider-component, or it will hide the range slider from the screenreader - exactly the opposite of what you want to do!

**Comment by NightCatSama** [comment]
I get it, I will support this feature. 

thank for your issue, this will be a meaningful update.

**Comment by callumacrae** [comment]
amazing, thank you!

**Comment by NightCatSama** [comment]
I added some accessibility attributes, but I do not have a device to test, so can you please help me test it.

I did not update the new version because I did not know if it was available.

demo: [https://nightcatsama.github.io/vue-slider-component/example/](https://nightcatsama.github.io/vue-slider-component/example/)

this is my references [ant's slider](https://ant.design/components/slider/).

**Comment by callumacrae** [comment]
Still doesn't work, because the screenreader doesn't know it's a slider. trying setting `role="slider"`: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_slider_role

**Comment by NightCatSama** [comment]
This attribute exists.

```html
<div 
  //...
  role="slider"
  :aria-valuemin="min"
  :aria-valuemax="max"
  :aria-valuenow="val"
  aria-disabled="false"
  aria-hidden="false"
>
..
</div>
```

**Comment by callumacrae** [comment]
I'm not able to change it using a screen reader :(

**Comment by NightCatSama** [comment]
Can you tell me whether the slider on the ant can be used? [slider](https://ant.design/components/slider/)

**Comment by callumacrae** [comment]
Nope, unable to use that with VoiceOver on Safari.

**Comment by NightCatSama** [comment]
It may still have to use the `<input type="range">`, but I don't  know how to hide the `input` tag in the common equipment.

**Comment by callumacrae** [comment]
See the article I linked to before: http://www.coolfields.co.uk/2016/05/text-for-screen-readers-only-updated/

```
.screen-reader-text { 
   clip: rect(1px, 1px, 1px, 1px); 
   height: 1px; 
   width: 1px; 
   overflow: hidden; 
   position: absolute !important;
}
```

stuff like that is pretty bulletproof

**Comment by NightCatSama** [comment]
I thought that this applies only to the text.

I'll give it a try.

**Comment by callumacrae** [comment]
nah, works for everything - it's more commonly implemented as `.sr-only`

**Comment by NightCatSama** [comment]
I have already corrected. [demo](https://nightcatsama.github.io/vue-slider-component/example/#demo)

Then I found that there is a problem with this scheme, not suitable for the range mode and custom data mode.

**Comment by callumacrae** [comment]
huh? not sure i follow

**Comment by NightCatSama** [comment]
Don't work?

**Comment by callumacrae** [comment]
Oh, right, I misunderstood.

That works, mostly! It reads out the existing value and it's possible to change it. The only problem is that the visually displayed slider, while the number displayed changes, the position of the handle does not.

**Comment by NightCatSama** [comment]
Do you mean that the slider position does not move when the value of input changes?

**Comment by callumacrae** [comment]
Yes, exactly. The value has changed, but it isn't reflected in the UI. Not a huge issue as it won't really affect anyone (the people using screenreaders generally won't see the UI anyway), but worth fixing if it's an easy fix.

**Comment by NightCatSama** [comment]
I do not know where the problem is, and it's hard for me to debug.

**Comment by callumacrae** [comment]
¯\_(ツ)_/¯ 

I think what you've done is enough - it works for people using screen readers now. Good job, thanks!

**Comment by spatras** [comment]
Hello,
I see that the slider doesn't support the screen reader. It has aria-hidden=true. Do you intend to change this? Or at least to give the option to change the aria-hidden value?
Thanks

**Comment by NightCatSama** [comment]
@spatras 

So `aria-hidden` can't be set to `true`? I don't understand very well, I wrote it according to the previous version (v2).

https://github.com/NightCatSama/vue-slider-component/blob/bc68cbe35e97e015036aeefcac6179c559fdf3f9/src/vue2-slider.vue#L9

I can't debug on the screen reader, if you can, please mention pr perfect components. Thanks

**Comment by spatras** [comment]
`aria-hidden` si set to `true`by default. I would like to be customisable to set it `false`. If I add `aria-valuemax`, and other aria properties on the handler, it doesn't work becasue the parent has `aria-hidden` `true`. It would be great if you could add a prop for `aria-hidden` value.
Thanks!

**Comment by NightCatSama** [comment]
@spatras 

So is aria-hidden= "true" a remained problems?  If so, then I set it to false by default to have other effects.

**Comment by spatras** [comment]
Yes. `aria-hidden="true"` is hiding the element and all it's childrens from the screen reader. You could remove the attribute entierly.

**Comment by NightCatSama** [comment]
@spatras Cancel this attribute in `v3.0.34`, and support adding custom attributes directly.

**Comment by spatras** [comment]
Awesome! Thanks!

**Comment by spatras** [comment]
Hi again!
Your last update helped a lot and many thanks for that but I found that there is an hidden input `.vue-slider-sr-only` which appear only when there is 1 slider handle and is read by the screen reader but can't interact with it. Can you remove that?

**Comment by NightCatSama** [comment]
@spatras  This is the code that was previously contributed by others. Can it be compatible with screen readers if it is removed?

**Comment by spatras** [comment]
@NightCatSama That code doesn't make the slider screen reader compatible. It only appears when there is only 1 handle and also, it cannot be changed from the screen reader. Maybe was doing more previously, but right doesn't do anything. And previously, when you had `aria-hidden` it was not read by the sr. For example users can add `aria-valuenow`, `aria-label`... and so on, to a custom handle and then you achieve fully sr compatibility.

**Comment by NightCatSama** [comment]
@spatras  Sorry, I don't know much about the compatibility of screen readers. Is it only necessary to delete `<input class="vue-slider-sr-only" />`?

**Comment by spatras** [comment]
Yes. Just that.

**Comment by NightCatSama** [comment]
@spatras Already removed, please update to v3.0.35.

**Comment by spatras** [comment]
Works great! Thank you!
